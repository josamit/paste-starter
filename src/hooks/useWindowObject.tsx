type UseWindowObjectReturn = {
  environment: string;
  username: string;
  email: string;
};

function useWindowObject(): UseWindowObjectReturn {
  return {
    // @ts-ignore
    environment: window?.environment || "stage",
    // @ts-ignore
    username: window?.user?.name || "jdoe",
    // @ts-ignore
    email: window?.user?.email || "jdoe@test.com",
  };
}

export { useWindowObject };
