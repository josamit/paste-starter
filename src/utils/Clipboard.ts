import { UseToasterReturnedProps } from "@twilio-paste/core";

const handleErr = (
  toaster: UseToasterReturnedProps,
  itemName: string,
  err: unknown
) => {
  toaster.push({
    message: `Oops, unable to copy ${itemName} to clipboard. ${err}`,
    variant: "error",
  });
};

const handleSuccess = (toaster: UseToasterReturnedProps, message: string) => {
  toaster.push({
    message: `${message}`,
    variant: "success",
  });
};

const handleWarning = (toaster: UseToasterReturnedProps, message: string) => {
  toaster.push({
    message: `${message}`,
    variant: "warning",
  });
};

export function copyToClipboardSync(
  textToCopy: string,
  itemName: string,
  toaster: UseToasterReturnedProps
) {
  const textArea = document.createElement("textarea");
  textArea.value = textToCopy;

  document.body.appendChild(textArea);
  textArea.focus();
  textArea.select();

  try {
    const copyResult = document.execCommand("copy");
    const message = copyResult
      ? `${itemName} was successfully copied to your clipboard.`
      : `${itemName} could not be copied to your clipboard. Please try again.`;
    if (copyResult) {
      handleSuccess(toaster, message);
    } else {
      handleWarning(toaster, message);
    }
  } catch (err) {
    handleErr(toaster, itemName, err);
  }

  document.body.removeChild(textArea);
}

export function copyToClipboardAsync(
  textToCopy: string,
  itemName: string,
  toaster: UseToasterReturnedProps
) {
  if (!navigator.clipboard) {
    copyToClipboardSync(textToCopy, itemName, toaster);
    return;
  }

  navigator.clipboard
    .writeText(textToCopy)
    .then(() =>
      handleSuccess(
        toaster,
        `${itemName} was successfully copied to your clipboard.`
      )
    )
    .catch((err) => handleErr(toaster, itemName, err));
}
