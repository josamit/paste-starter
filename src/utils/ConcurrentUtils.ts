/*
 * takes an array of items and a function that accepts an item at a time and returns a promise
 * returns a promise that resolves to an array of results
 */
export function mapConcurrent<T>(
  items: string | any[],
  maxConcurrent: number,
  fn: (arg0: T, arg1: number) => Promise<any>
) {
  let index = 0;
  let currentlyInFlightRequests = 0;
  let completedRequests = 0;
  const results = new Array(items.length);
  let stopExecution = false;

  return new Promise(function (resolve, reject) {
    function run() {
      // launch as many as we're allowed to
      while (
        !stopExecution &&
        currentlyInFlightRequests < maxConcurrent &&
        index < items.length
      ) {
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        runNext();
      }
      // if all are done, then resolve parent promise with results
      if (completedRequests === items.length) {
        resolve(results);
      }
    }

    function runNext() {
      const i = index;
      // eslint-disable-next-line no-plusplus
      ++currentlyInFlightRequests;
      // eslint-disable-next-line no-plusplus
      fn(items[index], index++).then(
        function (val) {
          // eslint-disable-next-line no-plusplus
          ++completedRequests;
          // eslint-disable-next-line no-plusplus
          --currentlyInFlightRequests;
          results[i] = val;
          run();
        },
        function (err) {
          // set flag so we don't launch any more requests
          stopExecution = true;
          reject(err);
        }
      );
    }

    run();
  });
}
