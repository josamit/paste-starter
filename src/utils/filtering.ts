import qs from "qs";
import { To } from "history";
import { QueryParams } from "../context/QueryManagerContext";
import { partiallyDecodeURI, partiallyEncodeURI } from "./url";

export function emptyQueryParams(push: (to: To, state?: any) => void) {
  push({ search: "" });
}

export function parseQueryParamsFromQueryString(queryString: string) {
  return qs.parse(queryString, {
    ignoreQueryPrefix: true,
    decoder: partiallyDecodeURI,
  });
}

export function parseQueryString(queryString: string): QueryParams {
  return parseQueryParamsFromQueryString(queryString);
}

export function stringifyWithEncoder(params: Record<string, string>) {
  return qs.stringify(params, {
    encoder: partiallyEncodeURI,
  });
}

export function setQueryParams(
  push: (to: To, state?: any) => void,
  pathname: string,
  params: QueryParams
): void {
  push({
    pathname,
    search: qs.stringify(params),
  });
}
