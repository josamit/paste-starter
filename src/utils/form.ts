import { fireInputChange } from "@twilio/react-typed-forms/dist/testing-utils";
import { fireEvent, RenderResult } from "@testing-library/react";

export function createGetForm({ findByTestId, queryByTestId }: RenderResult) {
  return (formId: string) => ({
    async get(testId: string) {
      return findByTestId(`form-${formId}-${testId}`);
    },
    async click(testId: string) {
      const element = await this.get(testId);

      fireEvent.click(element);
    },
    async type(testId: string, value: string) {
      const element = await this.get(testId);

      fireInputChange(element, value);
    },
    async submit() {
      const element = await this.get("submit-button");

      await fireEvent.click(element);
    },
    async expectValid() {
      const elementId = `${formId}-error`;

      expect(queryByTestId(elementId)).toBeNull();
    },
    async expectValue(testId: string, value: string) {
      const element = (await this.get(testId)) as HTMLInputElement;

      expect(element.value).toEqual(value);
    },
    async select(testId: string, value: string) {
      const element = await this.get(testId);

      await fireEvent.change(element, {
        target: {
          value,
        },
      });
    },
  });
}
