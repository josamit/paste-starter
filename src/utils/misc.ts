import { v4 as uuidv4 } from "uuid";

export function joinClassNames(...classNames: (string | undefined)[]): string {
  return classNames.filter(Boolean).join(" ");
}

// tslint:disable-next-line:no-empty
export const noop = () => {};

export function generateUniqueId() {
  return uuidv4();
}

type TypedEntry<T> = [keyof T, T[keyof T]];

export function typedEntries<T extends object>(obj: T): TypedEntry<T>[] {
  return Object.entries(obj) as TypedEntry<T>[];
}

export function typedKeys<T extends {}>(obj: T): (keyof T)[] {
  return Object.keys(obj) as (keyof T)[];
}

export function safelyParseDate(date?: string | null, asDate = false): string {
  if (date === undefined || date === null || date.trim().length === 0) {
    return "-";
  }
  try {
    const parsedDate = new Date(Date.parse(date));
    if (asDate) {
      return parsedDate.toLocaleDateString();
    }

    return parsedDate.toLocaleString();
  } catch (e) {
    return date;
  }
}
