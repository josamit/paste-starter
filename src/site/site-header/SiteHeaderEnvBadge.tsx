import React from "react";
import { Badge } from "@twilio-paste/core";
import { BadgeVariants } from "@twilio-paste/badge/dist/types";
import { useWindowObject } from "@hooks/useWindowObject";

const determineBadgeVariant = (environment: string) => {
  let variant: BadgeVariants = "error";

  if (environment === "stage") {
    variant = "warning";
  }

  if (environment === "dev") {
    variant = "success";
  }

  return variant;
};

const SiteHeaderEnvBadge: React.FC = () => {
  const { environment } = useWindowObject();

  const variant = determineBadgeVariant(environment);

  return (
    <Badge as="span" variant={variant}>
      {environment}
    </Badge>
  );
};

export default SiteHeaderEnvBadge;
