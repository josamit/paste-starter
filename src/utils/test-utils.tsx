import * as React from "react";
import { render as rtlRender } from "@testing-library/react";
import { Theme } from "@twilio-paste/core/theme";
import { MockQueryManagerContext } from "@context/QueryManagerContext";
import { WindowSizeContext } from "@context/WindowSizeProvider";
import { MemoryRouter, Route, Switch } from "react-router-dom";
import { DarkModeContext } from "@context/DarkModeContext";
import ToastsProvider from "@context/ToastsProvider";
import { createGetForm } from "@src/utils/form";
import { Box } from "@twilio-paste/core/dist/box";

export function renderWithFormUtils(...args: Parameters<typeof rtlRender>) {
  const [children, ...restArgs] = args;

  const results = rtlRender(<Box>{children}</Box>, ...restArgs);
  const getForm = createGetForm(results);

  return {
    ...results,
    getForm,
  };
}

const renderComponent = (
  ui: React.ReactElement,
  {
    initialEntry = "/route/1",
    initialRoute = "/route/:appSid",
    queryParams = {
      accountSid: "ACxxx",
      activeTab: "a2pBrandsReconciliation",
    },
    container = undefined,
    mockToggleMode = jest.fn(),
    setQueryParams = jest.fn(),
    ...renderOptions
  } = {}
) => {
  const Wrapper: React.FC = ({ children }) => (
    <MockQueryManagerContext
      value={{
        queryParams,
        setQueryParams,
      }}
    >
      <WindowSizeContext.Provider
        value={{ width: 1400, height: 800, breakpointIndex: 450 }}
      >
        <Theme.Provider theme="default" disableAnimations>
          <DarkModeContext.Provider
            value={{
              theme: "default",
              toggleMode: mockToggleMode,
              componentMounted: true,
            }}
          >
            <ToastsProvider>
              <MemoryRouter initialEntries={[initialEntry]}>
                <Switch>
                  <Route path={initialRoute}>{children}</Route>
                </Switch>
              </MemoryRouter>
            </ToastsProvider>
          </DarkModeContext.Provider>
        </Theme.Provider>
      </WindowSizeContext.Provider>
    </MockQueryManagerContext>
  );

  return renderWithFormUtils(ui, {
    wrapper: Wrapper,
    container,
    ...renderOptions,
  });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { renderComponent };
