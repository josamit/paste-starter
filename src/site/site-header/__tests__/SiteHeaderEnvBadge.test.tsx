import * as React from "react";
import { renderComponent } from "@src/utils/test-utils";
import SiteHeaderEnvBadge from "@site/site-header/SiteHeaderEnvBadge";
import * as useWindowObjectHook from "@hooks/useWindowObject";

describe("<SiteHeaderEnvBadge />", () => {
  it("should render correctly for dev env", async () => {
    jest
      .spyOn(useWindowObjectHook, "useWindowObject")
      .mockImplementation(() => ({
        environment: "dev",
        username: "test",
        email: "test@test.com",
      }));

    const { container } = renderComponent(<SiteHeaderEnvBadge />);

    expect(container).toMatchSnapshot();
  });
  it("should render correctly for stage env", async () => {
    jest
      .spyOn(useWindowObjectHook, "useWindowObject")
      .mockImplementation(() => ({
        environment: "stage",
        username: "test",
        email: "test@test.com",
      }));

    const { container } = renderComponent(<SiteHeaderEnvBadge />);

    expect(container).toMatchSnapshot();
  });
  it("should render correctly for prod env", async () => {
    jest
      .spyOn(useWindowObjectHook, "useWindowObject")
      .mockImplementation(() => ({
        environment: "prod",
        username: "test",
        email: "test@test.com",
      }));

    const { container } = renderComponent(<SiteHeaderEnvBadge />);

    expect(container).toMatchSnapshot();
  });
});
