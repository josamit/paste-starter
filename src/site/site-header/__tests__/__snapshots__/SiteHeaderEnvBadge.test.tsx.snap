// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SiteHeaderEnvBadge /> should render correctly for dev env 1`] = `
<div>
  <div
    class="css-ql39e"
  >
    <div
      class="css-roynbj"
      data-paste-element="BOX"
    >
      <span
        class="css-psiz1w"
        color="colorTextSuccess"
        data-paste-element="BADGE"
        display="flex"
        font-size="fontSize20"
        font-weight="fontWeightSemibold"
      >
        dev
      </span>
    </div>
  </div>
</div>
`;

exports[`<SiteHeaderEnvBadge /> should render correctly for prod env 1`] = `
<div>
  <div
    class="css-ql39e"
  >
    <div
      class="css-roynbj"
      data-paste-element="BOX"
    >
      <span
        class="css-1php4p5"
        color="colorTextError"
        data-paste-element="BADGE"
        display="flex"
        font-size="fontSize20"
        font-weight="fontWeightSemibold"
      >
        prod
      </span>
    </div>
  </div>
</div>
`;

exports[`<SiteHeaderEnvBadge /> should render correctly for stage env 1`] = `
<div>
  <div
    class="css-ql39e"
  >
    <div
      class="css-roynbj"
      data-paste-element="BOX"
    >
      <span
        class="css-thuikx"
        color="colorTextWarningStrong"
        data-paste-element="BADGE"
        display="flex"
        font-size="fontSize20"
        font-weight="fontWeightSemibold"
      >
        stage
      </span>
    </div>
  </div>
</div>
`;
